src/dc_mimeparser.rs

```diff
@@ -1277,35 +1227,42 @@ pub unsafe fn mailmime_transfer_decode(
         || mime_transfer_encoding == MAILMIME_MECHANISM_8BIT as libc::c_int
         || mime_transfer_encoding == MAILMIME_MECHANISM_BINARY as libc::c_int
     {
-        decoded_data = (*mime_data).dt_data.dt_text.dt_data;
-        decoded_data_bytes = (*mime_data).dt_data.dt_text.dt_length;
+        let decoded_data = (*mime_data).dt_data.dt_text.dt_data;
+        let decoded_data_bytes = (*mime_data).dt_data.dt_text.dt_length;
+
         if decoded_data.is_null() || decoded_data_bytes <= 0 {
-            return false;
-        }
-    } else {
-        let mut current_index = 0;
-        let r = mailmime_part_parse(
-            (*mime_data).dt_data.dt_text.dt_data,
-            (*mime_data).dt_data.dt_text.dt_length,
-            &mut current_index,
-            mime_transfer_encoding,
-            &mut transfer_decoding_buffer,
-            &mut decoded_data_bytes,
-        );
-        if r != MAILIMF_NO_ERROR as libc::c_int
-            || transfer_decoding_buffer.is_null()
-            || decoded_data_bytes <= 0
-        {
-            return false;
+            bail!("No data to decode found");
+        } else {
+            let result = std::slice::from_raw_parts(decoded_data as *const u8, decoded_data_bytes);
+            return Ok(result.to_vec());
         }
-        decoded_data = transfer_decoding_buffer;
     }
 
-    *ret_decoded_data = decoded_data;
-    *ret_decoded_data_bytes = decoded_data_bytes;
-    *ret_to_mmap_string_unref = transfer_decoding_buffer;
+    let mut current_index = 0;
+    let mut transfer_decoding_buffer = ptr::null_mut();
+    let mut decoded_data_bytes = 0;
+
+    let r = mailmime_part_parse(
+        (*mime_data).dt_data.dt_text.dt_data,
+        (*mime_data).dt_data.dt_text.dt_length,
+        &mut current_index,
+        mime_transfer_encoding,
+        &mut transfer_decoding_buffer,
+        &mut decoded_data_bytes,
+    );
+
+    if r == MAILIMF_NO_ERROR as libc::c_int
+        && !transfer_decoding_buffer.is_null()
+        && decoded_data_bytes > 0
+    {
+        let result =
+            std::slice::from_raw_parts(transfer_decoding_buffer as *const u8, decoded_data_bytes);
+        mmap_string_unref(transfer_decoding_buffer);
+
+        return Ok(result.to_vec());
```

Comment:
huh? doesn't result become dangling after unref()?

Reply:
damn it, I am getting sloppy

Reply:
fixed